File: pengepul/oauth.py
from __future__ import annotations

import asyncio
from urllib.parse import quote, urlencode

import httpx

from .types import PKCECodes, RefreshTokenExhaustedError, TokenData
from .utils import decode_jwt_payload, expires_in_iso

ANTHROPIC_AUTH_URL = "https://claude.ai/oauth/authorize"
ANTHROPIC_TOKEN_URL = "https://api.anthropic.com/v1/oauth/token"
ANTHROPIC_CLIENT_ID = "9d1c250a-e61b-44d9-88ed-5944d1962f5e"
ANTHROPIC_REDIRECT_URI = "http://localhost:54545/callback"
ANTHROPIC_SCOPE = "org:create_api_key user:profile user:inference"

CODEX_ISSUER = "https://auth.openai.com"
CODEX_AUTH_URL = f"{CODEX_ISSUER}/oauth/authorize"
CODEX_TOKEN_URL = f"{CODEX_ISSUER}/oauth/token"
CODEX_CLIENT_ID = "app_EMoamEEZ73f0CkXaXp7hrann"
CODEX_CALLBACK_PORT = 1455
CODEX_CALLBACK_PATH = "/auth/callback"
CODEX_REDIRECT_URI = f"http://localhost:{CODEX_CALLBACK_PORT}{CODEX_CALLBACK_PATH}"
CODEX_SCOPE = "openid profile email offline_access api.connectors.read api.connectors.invoke"
CODEX_ORIGINATOR = "codex_cli_rs"


def detect_exhausted_reason(body: str) -> str | None:
    lowered = body.lower()
    for marker in (
        "refresh_token_reused",
        "invalid_grant",
        "expired",
        "invalidated",
        "revoked",
    ):
        if marker in lowered:
            return marker
    return None


def generate_anthropic_auth_url(state: str, pkce: PKCECodes) -> str:
    params = urlencode(
        {
            "code": "true",
            "client_id": ANTHROPIC_CLIENT_ID,
            "response_type": "code",
            "redirect_uri": ANTHROPIC_REDIRECT_URI,
            "code_challenge": pkce.code_challenge,
            "code_challenge_method": "S256",
            "state": state,
        }
    )
    scope = "+".join(quote(part, safe=":") for part in ANTHROPIC_SCOPE.split())
    return f"{ANTHROPIC_AUTH_URL}?{params}&scope={scope}"


def generate_codex_auth_url(state: str, pkce: PKCECodes) -> str:
    params = urlencode(
        {
            "response_type": "code",
            "client_id": CODEX_CLIENT_ID,
            "redirect_uri": CODEX_REDIRECT_URI,
            "scope": CODEX_SCOPE,
            "code_challenge": pkce.code_challenge,
            "code_challenge_method": "S256",
            "id_token_add_organizations": "true",
            "codex_cli_simplified_flow": "true",
            "state": state,
            "originator": CODEX_ORIGINATOR,
        }
    )
    return f"{CODEX_AUTH_URL}?{params}"


def _raise_state_mismatch(returned_state: str, expected_state: str) -> None:
    if returned_state != expected_state:
        raise ValueError("OAuth state mismatch")


def _anthropic_token(data: dict) -> TokenData:
    account = data.get("account") or {}
    return TokenData(
        access_token=data["access_token"],
        refresh_token=data["refresh_token"],
        email=account.get("email_address") or "unknown",
        expires_at=expires_in_iso(data.get("expires_in")),
        account_uuid=account.get("uuid") or "",
        provider="anthropic",
    )


def _codex_identity(id_token: str) -> tuple[str, str, str | None]:
    claims = decode_jwt_payload(id_token)
    auth = claims.get("https://api.openai.com/auth") or {}
    email = claims.get("email") or "unknown"
    account_id = ""
    plan_type = None
    if isinstance(auth, dict):
        account_id = auth.get("chatgpt_account_id") or ""
        plan_type = auth.get("chatgpt_plan_type")
    account_id = account_id or claims.get("chatgpt_account_id") or ""
    plan_type = plan_type or claims.get("chatgpt_plan_type")
    return str(email), str(account_id), str(plan_type) if plan_type else None


def _codex_token(data: dict) -> TokenData:
    id_token = data["id_token"]
    email, account_id, plan_type = _codex_identity(id_token)
    return TokenData(
        access_token=data["access_token"],
        refresh_token=data["refresh_token"],
        email=email,
        expires_at=expires_in_iso(data.get("expires_in"), default_seconds=3600),
        account_uuid=account_id,
        provider="codex",
        id_token=id_token,
        plan_type=plan_type,
    )


async def exchange_anthropic_code(
    code: str,
    returned_state: str,
    expected_state: str,
    pkce: PKCECodes,
) -> TokenData:
    _raise_state_mismatch(returned_state, expected_state)
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            ANTHROPIC_TOKEN_URL,
            json={
                "code": code,
                "grant_type": "authorization_code",
                "client_id": ANTHROPIC_CLIENT_ID,
                "redirect_uri": ANTHROPIC_REDIRECT_URI,
                "code_verifier": pkce.code_verifier,
                "state": expected_state,
            },
        )
    if resp.status_code >= 400:
        raise RuntimeError(f"token exchange failed ({resp.status_code}): {resp.text}")
    return _anthropic_token(resp.json())


async def refresh_anthropic_tokens(refresh_token: str) -> TokenData:
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            ANTHROPIC_TOKEN_URL,
            json={
                "client_id": ANTHROPIC_CLIENT_ID,
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
            },
        )
    if resp.status_code >= 400:
        reason = detect_exhausted_reason(resp.text)
        if reason:
            raise RefreshTokenExhaustedError(reason, resp.status_code, resp.text)
        raise RuntimeError(f"token refresh failed ({resp.status_code}): {resp.text}")
    return _anthropic_token(resp.json())


async def exchange_codex_code(
    code: str,
    returned_state: str,
    expected_state: str,
    pkce: PKCECodes,
) -> TokenData:
    _raise_state_mismatch(returned_state, expected_state)
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            CODEX_TOKEN_URL,
            data={
                "grant_type": "authorization_code",
                "code": code,
                "redirect_uri": CODEX_REDIRECT_URI,
                "client_id": CODEX_CLIENT_ID,
                "code_verifier": pkce.code_verifier,
            },
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )
    if resp.status_code >= 400:
        raise RuntimeError(f"codex token exchange failed ({resp.status_code}): {resp.text}")
    return _codex_token(resp.json())


async def refresh_codex_tokens(refresh_token: str) -> TokenData:
    async with httpx.AsyncClient(timeout=30) as client:
        resp = await client.post(
            CODEX_TOKEN_URL,
            json={
                "client_id": CODEX_CLIENT_ID,
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
            },
        )
    if resp.status_code >= 400:
        reason = detect_exhausted_reason(resp.text)
        if reason:
            raise RefreshTokenExhaustedError(reason, resp.status_code, resp.text)
        raise RuntimeError(f"codex token refresh failed ({resp.status_code}): {resp.text}")
    return _codex_token(resp.json())


async def refresh_with_retry(refresh_token: str, provider: str, max_retries: int = 3) -> TokenData:
    last_error: BaseException | None = None
    for attempt in range(1, max_retries + 1):
        try:
            if provider == "codex":
                return await refresh_codex_tokens(refresh_token)
            return await refresh_anthropic_tokens(refresh_token)
        except RefreshTokenExhaustedError:
            raise
        except Exception as exc:
            last_error = exc
            if attempt == max_retries:
                break
            await asyncio.sleep(attempt)
    assert last_error is not None
    raise last_error

File: src/oauth.rs
use url::form_urlencoded::Serializer;

use anyhow::{Context, Result, bail};
use serde_json::Value;

use crate::types::{PkceCodes, ProviderId, RefreshTokenExhaustedError, TokenData};
use crate::utils::{decode_jwt_payload, expires_in_iso};

pub const ANTHROPIC_AUTH_URL: &str = "https://claude.ai/oauth/authorize";
pub const ANTHROPIC_CLIENT_ID: &str = "9d1c250a-e61b-44d9-88ed-5944d1962f5e";
pub const ANTHROPIC_REDIRECT_URI: &str = "http://localhost:54545/callback";
pub const ANTHROPIC_SCOPE: &str = "org:create_api_key user:profile user:inference";

pub const CODEX_ISSUER: &str = "https://auth.openai.com";
pub const CODEX_AUTH_URL: &str = "https://auth.openai.com/oauth/authorize";
pub const CODEX_CLIENT_ID: &str = "app_EMoamEEZ73f0CkXaXp7hrann";
pub const CODEX_CALLBACK_PORT: u16 = 1455;
pub const CODEX_CALLBACK_PATH: &str = "/auth/callback";
pub const CODEX_SCOPE: &str =
    "openid profile email offline_access api.connectors.read api.connectors.invoke";
pub const CODEX_ORIGINATOR: &str = "codex_cli_rs";
pub const ANTHROPIC_TOKEN_URL: &str = "https://api.anthropic.com/v1/oauth/token";
pub const CODEX_TOKEN_URL: &str = "https://auth.openai.com/oauth/token";

#[must_use]
pub fn detect_exhausted_reason(body: &str) -> Option<&'static str> {
    let body = body.to_ascii_lowercase();
    [
        "refresh_token_reused",
        "invalid_grant",
        "expired",
        "invalidated",
        "revoked",
    ]
    .into_iter()
    .find(|marker| body.contains(marker))
}

#[must_use]
pub fn generate_anthropic_auth_url(state: &str, pkce: &PkceCodes) -> String {
    let query = Serializer::new(String::new())
        .append_pair("code", "true")
        .append_pair("client_id", ANTHROPIC_CLIENT_ID)
        .append_pair("response_type", "code")
        .append_pair("redirect_uri", ANTHROPIC_REDIRECT_URI)
        .append_pair("code_challenge", &pkce.code_challenge)
        .append_pair("code_challenge_method", "S256")
        .append_pair("state", state)
        .append_pair("scope", ANTHROPIC_SCOPE)
        .finish();
    format!("{ANTHROPIC_AUTH_URL}?{query}")
}

#[must_use]
pub fn generate_codex_auth_url(state: &str, pkce: &PkceCodes) -> String {
    let redirect_uri = format!("http://localhost:{CODEX_CALLBACK_PORT}{CODEX_CALLBACK_PATH}");
    let query = Serializer::new(String::new())
        .append_pair("response_type", "code")
        .append_pair("client_id", CODEX_CLIENT_ID)
        .append_pair("redirect_uri", &redirect_uri)
        .append_pair("scope", CODEX_SCOPE)
        .append_pair("code_challenge", &pkce.code_challenge)
        .append_pair("code_challenge_method", "S256")
        .append_pair("id_token_add_organizations", "true")
        .append_pair("codex_cli_simplified_flow", "true")
        .append_pair("state", state)
        .append_pair("originator", CODEX_ORIGINATOR)
        .finish();
    format!("{CODEX_AUTH_URL}?{query}")
}

/// Exchange an Anthropic OAuth authorization code for a stored token.
///
/// # Errors
///
/// Returns an error when OAuth state does not match, the token endpoint fails, or the response
/// body does not contain the expected token fields.
pub async fn exchange_anthropic_code(
    code: &str,
    returned_state: &str,
    expected_state: &str,
    pkce: &PkceCodes,
) -> Result<TokenData> {
    ensure_state(returned_state, expected_state)?;
    let response = reqwest::Client::new()
        .post(ANTHROPIC_TOKEN_URL)
        .json(&serde_json::json!({
            "code": code,
            "grant_type": "authorization_code",
            "client_id": ANTHROPIC_CLIENT_ID,
            "redirect_uri": ANTHROPIC_REDIRECT_URI,
            "code_verifier": pkce.code_verifier,
            "state": expected_state
        }))
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await?;
    let status = response.status();
    let body = response.text().await?;
    if !status.is_success() {
        bail!("token exchange failed ({status}): {body}");
    }
    anthropic_token(&serde_json::from_str(&body).context("Anthropic token response is not JSON")?)
}

/// Build the Anthropic OAuth refresh request (JSON body).
fn anthropic_refresh_request(refresh_token: &str) -> reqwest::RequestBuilder {
    reqwest::Client::new()
        .post(ANTHROPIC_TOKEN_URL)
        .json(&serde_json::json!({
            "client_id": ANTHROPIC_CLIENT_ID,
            "grant_type": "refresh_token",
            "refresh_token": refresh_token,
        }))
        .timeout(std::time::Duration::from_secs(30))
}

/// Refresh an Anthropic OAuth token.
///
/// # Errors
///
/// Returns an error when the token endpoint fails or the response body is invalid.
pub async fn refresh_anthropic_tokens(refresh_token: String) -> Result<TokenData> {
    let response = anthropic_refresh_request(&refresh_token).send().await?;
    let status = response.status();
    let body = response.text().await?;
    if !status.is_success() {
        if let Some(reason) = detect_exhausted_reason(&body) {
            return Err(
                RefreshTokenExhaustedError::new(reason, Some(status.as_u16()), Some(body)).into(),
            );
        }
        bail!("token refresh failed ({status}): {body}");
    }
    anthropic_token(&serde_json::from_str(&body).context("Anthropic refresh response is not JSON")?)
}

/// Exchange a Codex OAuth authorization code for a stored token.
///
/// # Errors
///
/// Returns an error when OAuth state does not match, the token endpoint fails, or the response
/// body does not contain the expected token fields.
pub async fn exchange_codex_code(
    code: &str,
    returned_state: &str,
    expected_state: &str,
    pkce: &PkceCodes,
) -> Result<TokenData> {
    ensure_state(returned_state, expected_state)?;
    let redirect_uri = format!("http://localhost:{CODEX_CALLBACK_PORT}{CODEX_CALLBACK_PATH}");
    let response = reqwest::Client::new()
        .post(CODEX_TOKEN_URL)
        .form(&[
            ("grant_type", "authorization_code"),
            ("code", code),
            ("redirect_uri", redirect_uri.as_str()),
            ("client_id", CODEX_CLIENT_ID),
            ("code_verifier", pkce.code_verifier.as_str()),
        ])
        .timeout(std::time::Duration::from_secs(30))
        .send()
        .await?;
    let status = response.status();
    let body = response.text().await?;
    if !status.is_success() {
        bail!("codex token exchange failed ({status}): {body}");
    }
    codex_token(&serde_json::from_str(&body).context("Codex token response is not JSON")?)
}

/// Build the Codex OAuth refresh request (form-encoded body).
fn codex_refresh_request(refresh_token: &str) -> reqwest::RequestBuilder {
    reqwest::Client::new()
        .post(CODEX_TOKEN_URL)
        .form(&[
            ("grant_type", "refresh_token"),
            ("client_id", CODEX_CLIENT_ID),
            ("refresh_token", refresh_token),
        ])
        .timeout(std::time::Duration::from_secs(30))
}

/// Refresh a Codex OAuth token.
///
/// # Errors
///
/// Returns an error when the token endpoint fails or the response body is invalid.
pub async fn refresh_codex_tokens(refresh_token: String) -> Result<TokenData> {
    let response = codex_refresh_request(&refresh_token).send().await?;
    let status = response.status();
    let body = response.text().await?;
    if !status.is_success() {
        if let Some(reason) = detect_exhausted_reason(&body) {
            return Err(
                RefreshTokenExhaustedError::new(reason, Some(status.as_u16()), Some(body)).into(),
            );
        }
        bail!("codex token refresh failed ({status}): {body}");
    }
    codex_token(&serde_json::from_str(&body).context("Codex refresh response is not JSON")?)
}

fn ensure_state(returned_state: &str, expected_state: &str) -> Result<()> {
    if returned_state != expected_state {
        bail!("OAuth state mismatch");
    }
    Ok(())
}

fn anthropic_token(data: &Value) -> Result<TokenData> {
    let account = data.get("account").unwrap_or(&Value::Null);
    Ok(TokenData {
        access_token: required_string(data, "access_token")?,
        refresh_token: required_string(data, "refresh_token")?,
        email: account
            .get("email_address")
            .and_then(Value::as_str)
            .unwrap_or("unknown")
            .to_string(),
        expires_at: expires_in_iso(data.get("expires_in").and_then(Value::as_u64), 3600),
        account_uuid: account
            .get("uuid")
            .and_then(Value::as_str)
            .unwrap_or("")
            .to_string(),
        provider: ProviderId::Anthropic,
        id_token: None,
        last_refresh_at: None,
        plan_type: None,
        cursor: None,
    })
}

fn codex_token(data: &Value) -> Result<TokenData> {
    let id_token = required_string(data, "id_token")?;
    let claims = decode_jwt_payload(&id_token)?;
    let auth = claims
        .get("https://api.openai.com/auth")
        .unwrap_or(&Value::Null);
    let email = claims
        .get("email")
        .and_then(Value::as_str)
        .unwrap_or("unknown")
        .to_string();
    let account_uuid = auth
        .get("chatgpt_account_id")
        .or_else(|| claims.get("chatgpt_account_id"))
        .and_then(Value::as_str)
        .unwrap_or("")
        .to_string();
    let plan_type = auth
        .get("chatgpt_plan_type")
        .or_else(|| claims.get("chatgpt_plan_type"))
        .and_then(Value::as_str)
        .map(ToOwned::to_owned);
    Ok(TokenData {
        access_token: required_string(data, "access_token")?,
        refresh_token: required_string(data, "refresh_token")?,
        email,
        expires_at: expires_in_iso(data.get("expires_in").and_then(Value::as_u64), 3600),
        account_uuid,
        provider: ProviderId::Codex,
        id_token: Some(id_token),
        last_refresh_at: None,
        plan_type,
        cursor: None,
    })
}

fn required_string(data: &Value, field: &str) -> Result<String> {
    data.get(field)
        .and_then(Value::as_str)
        .map(ToOwned::to_owned)
        .with_context(|| format!("token response is missing {field}"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn content_type(req: &reqwest::Request) -> String {
        req.headers()
            .get(reqwest::header::CONTENT_TYPE)
            .and_then(|value| value.to_str().ok())
            .unwrap_or_default()
            .to_owned()
    }

    fn body_string(req: &reqwest::Request) -> String {
        let bytes = req
            .body()
            .and_then(reqwest::Body::as_bytes)
            .expect("refresh request always sets a body");
        std::str::from_utf8(bytes)
            .expect("refresh request body is valid utf-8")
            .to_owned()
    }

    #[test]
    fn codex_refresh_uses_form_encoded_request() {
        let req = codex_refresh_request("refresh-xyz")
            .build()
            .expect("codex refresh request builds");
        assert_eq!(req.url().as_str(), CODEX_TOKEN_URL);
        assert_eq!(content_type(&req), "application/x-www-form-urlencoded");
        let body = body_string(&req);
        assert!(body.contains("grant_type=refresh_token"), "{body}");
        assert!(body.contains("refresh_token=refresh-xyz"), "{body}");
    }

    #[test]
    fn anthropic_refresh_uses_json_request() {
        let req = anthropic_refresh_request("refresh-xyz")
            .build()
            .expect("anthropic refresh request builds");
        assert_eq!(req.url().as_str(), ANTHROPIC_TOKEN_URL);
        assert_eq!(content_type(&req), "application/json");
        let body = body_string(&req);
        assert!(body.contains("\"grant_type\":\"refresh_token\""), "{body}");
        assert!(body.contains("\"refresh_token\":\"refresh-xyz\""), "{body}");
    }
}
